src/App.tsx

```diff
@@ -0,0 +1,795 @@
+    urlUtils.saveParamsToUrl(
+      collection || null,
+      datasetName,
+      featureName,
+      selectedTrack ? selectedTrack.trackId : null,
+      currentFrame
+    );
+  }, [collection, datasetName, featureName, selectedTrack, currentFrame]);
+
+  /**
+   * Handles updating the canvas from state (for some common properties) and initiating the canvas
+   * render. Also updates other UI elements, including the plot and URL.
+   */
+  const drawLoop = useCallback(async (): Promise<void> => {
+    // Note: Selected track, frame number, etc. are not updated here.
+    // Those operations are async, and need to complete before a state update to be
+    // rendered correctly.
+    canv.setShowTrackPath(showTrackPath);
+    canv.setHideValuesOutOfRange(hideValuesOutOfRange);
+    canv.setColorRamp(colorRamp);
+    canv.setColorMapRangeMin(colorRampMin);
+    canv.setColorMapRangeMax(colorRampMax);
+    canv.setSelectedTrack(selectedTrack);
+    await canv.render();
+
+    // update current time in plot
+    plot?.setTime(currentFrame);
+
+    if (!timeControls?.isPlaying() && !recordingControls?.isRecording()) {
+      // Do not update URL while playback is happening for performance + UX reasons
+      updateUrl();
+    }
+  }, [
+    dataset,
+    datasetName,
+    featureName,
+    currentFrame,
+    selectedTrack,
+    hideValuesOutOfRange,
+    showTrackPath,
+    colorRamp,
+    colorRampMin,
+    colorRampMax,
+    timeControls?.isPlaying(), // updates URL when timeControls stops
+    updateUrl,
+  ]);
+
+  // Call drawLoop whenever its dependencies change.
+  useMemo(() => {
+    drawLoop();
+  }, [drawLoop]);
+
+  const setFrame = useCallback(
+    async (frame: number) => {
+      await canv.setFrame(frame);
+      setCurrentFrame(frame);
+      setFrameInput(frame);
+    },
+    [drawLoop, canv]
+  );
+
+  const findTrack = useCallback(
+    async (trackId: number, seekToFrame: boolean = true): Promise<void> => {
+      const newTrack = dataset!.buildTrack(trackId);
+
+      if (newTrack.length() < 1) {
+        // Check track validity
+        return;
+      }
+      setSelectedTrack(newTrack);
+      if (seekToFrame) {
+        setFrame(newTrack.times[0]);
+        plot?.plot(newTrack, featureName, newTrack.times[0]);
+      } else {
+        plot?.plot(newTrack, featureName, currentFrame);
+      }
+      setFindTrackInput("" + trackId);
+      updateUrl();
+    },
+    [canv, plot, dataset, featureName, currentFrame]
+  );
+
+  /**
+   * Update the canvas dimensions based on the current window size.
+   */
+  const setSize = (): void => canv.setSize(Math.min(window.innerWidth, 730), Math.min(window.innerHeight, 500));
+
+  // INITIAL SETUP  ////////////////////////////////////////////////////////////////
+
+  // Memoized because the URL can update as data is loaded in
+  const initialUrlParams = useMemo(urlUtils.loadParamsFromUrl, []);
+
+  // Load database and collections data from the URL.
+  useMemo(async () => {
```

Comment:
should this have an explicit dependency on initialUrlParams?

Reply:
I think this is okay without the explicit dependency, since initialUrlParams will never change after the first call to `useMemo(urlUtils.loadParamsFromUrl)`, and because we only want the second `useMemo` that loads in the database/collections to run once and therefore it should have no dependencies.

I'm going to expand the comment here a bit to explain what it's doing more concretely, though.